### DISCUSSION TITLE
Support for Hetzner Cloud in Supabase

### DISCUSSION THREAD

**Discussion by nicokempe** [supabase/supabase#28230] [Feature Requests]
**Summary:**
Integrating native support for [Hetzner Cloud](https://www.hetzner.com/cloud/) into [Supabase](https://supabase.com/) would significantly benefit both the platform and its users by offering a cost-effective, reliable, and environmentally friendly alternative to the current cloud provider AWS.

**Benefits:**

1. **Cost Efficiency:**
   - **Lower Costs:** Hetzner offers significantly lower server prices compared to AWS, Azure, and Google Cloud, reducing operational costs for both Supabase and its customers. For example, a small Hetzner cloud instance costs only €3.17 per month with 20 TB traffic included, compared to $30.37 for a similar instance on AWS.
   - **Increased Margins:** Supabase can maintain or even increase profit margins due to Hetzner's lower operational costs while offering more competitive prices to customers.

2. **Enhanced Customer Attraction and Retention:**
   - **Affordable Dedicated Resources:** Hetzner's affordable dedicated IP addresses and other resources make it an attractive option for businesses of all sizes. Hetzner provides IPv4 addresses at €0.40 per month and free IPv6 addresses, which is significantly cheaper than AWS and Azure.
   - **High-Quality Service:** Hetzner's reputation for high reliability and performance ensures that customers receive top-tier service without the high costs associated with other providers. I have only had a relationship with Hetzner as a very satisfied customer for years. Both in a business and private context. I have already been with many hosters (AWS, Azure, Google Cloud, Nitrado, StartyourServer, MC-Host, DeinserverHost, Fee-Hosting, ...). But since I switched to Hetzner a few years ago, there's been no going back. Not a single noticeable outage, uptime of over a year, the support is great and the prices unbeatable. What's more, they're more environmentally friendly and it's easier to believe that they don't mess around with your data in terms of data protection.

3. **Environmental Benefits:**
   - **Sustainability:** Hetzner is known for its commitment to environmental sustainability, including comprehensive certifications for environmental and security standards. They utilize renewable energy sources in their data centers and have received multiple awards for their sustainability efforts. Partnering with such a provider aligns with the growing demand for eco-friendly tech solutions.
   - **Transparency:** Hetzner offers tours of their data centers, showcasing their sustainable practices and operational transparency, further bolstering customer trust and satisfaction. This transparency is exemplified by YouTuber Raspberry Pi Cloud's tour of Hetzner's data center.

4. **Simplified Pricing and Reduced Hidden Costs:**
   - **Transparent Pricing:** Hetzner's pricing model is straightforward and free of hidden costs. For example, they do not charge for incoming traffic, and outgoing traffic costs are minimal, with 20-60 TB included per month and only €1 per additional TB. This contrasts sharply with the complex and often opaque pricing structures of AWS and Azure, which can lead to unexpected expenses.

**Conclusion:**
Implementing native support for Hetzner Cloud would not only provide a more cost-effective solution for Supabase's current and potential customers but also enhance the platform's appeal through environmental sustainability, high-quality service, and transparent pricing. This strategic move could result in increased customer acquisition, retention, and overall satisfaction, benefiting both Supabase and its users.

**Notes:**
I have written an internal comparison of Hetzner, STACKIT, AWS, Google Cloud, and Azure a few months ago for a company I work for and I could provide some numbers if relevant. I really like using Supabase and I really like using Hetzner. Self-hosted is of course an option, but I'd rather be able to concentrate on my core business processes and throw my money at you so that you, the experts, can take care of it. I would be very grateful if you would at least seriously consider hosting with Hetzner. I'm sure you could really benefit from it.

**Comment by eqoram** [comment]
Something like: https://elest.io/open-source/supabase 
But I would also rather support Supabase directly

**Reply by nicokempe** [reply]
Yes, the Hetzner Cloud is super cheap, more sustainable, has much more transparent prices and is also easier to manage. The only downside is that they don't offer managed databases or Kubernetes. If they were to build on Hetzner Cloud and simply offer that as an option alongside AWS, they could halve their prices and still quadruple their margin. 

In addition, I attach great importance to data protection in a German company and feel much more comfortable with providers such as Hetzner and STACKIT.

**Comment by suryavirkapur** [comment]
What about users located in the Middle East, LATAM and India? We do not have Hetzner region available close to us.

**Reply by nicokempe** [reply]
Thanks for the feedback. The idea is to add Hetzner as an additional option, not to replace AWS or other providers. This would give users the flexibility to choose based on their specific needs, such as cost, location, or data privacy preferences.

Hetzner could be a strong choice for EU-based users or those looking for more affordable and sustainable options, while AWS and others can still serve regions like the Middle East, LATAM, and India. And also, Hetzner doesn't only have data centers in the EU; they also have locations in the US (Ashburn, VA, and Hillsboro, OR) and Singapore, which could cover some non-EU users' needs. More options would mean better choices for everyone.

Let me know if you have further thoughts.

**Reply by ChrisFeldmeier** [reply]
Hello, do you know a Hetzner Cloud Supabase setup / auto installer or anything else?

**Reply by nicokempe** [reply]
Not as far as I know, but there is [a article for manual self-hosting](https://blog.activeno.de/the-ultimate-supabase-self-hosting-guide). It is from end of 2023 so maybe a bit outdated with the docker-compose, but you can easily get that from the [official Supabase docs](https://supabase.com/docs/guides/self-hosting/docker).

Now that Hetzner Cloud does offer S3 production-ready you may want to use those instead of the way described in the article.

It would maybe be a good idea to ask Hetzner if they could provide a image for Supabase. Wouldn't achieve what I wanted to achieve with this issue (managed Supabase by Supabase itself just to get away from scaling problems in the long run) but that would definetly make a lot easier for people that don't want to spend too much time with Docker and initial Supabase configuration through the Docker Compose if they want to use a way just like they did for Grafana and other images.

For a visual walkthrough, consider this tutorial: https://youtu.be/lHfgnFmQ1Ds

**Comment by nietsmmar** [comment]
Any news on this one? This would be awesome for our EU-based company. Our clients in the education sector would love when our backend is hosted by a European company rather than Amazon.

**Reply by aantti** [reply]
I don't think there are any plans for the platform to switch from AWS. I'm testing [self-hosted Supabase](https://supabase.com/docs/guides/self-hosting) on Hetzner regularly though.

**Reply by nietsmmar** [reply]
@aantti How do you deploy it? Do you use self-hosting with docker? Or is there any guide you use when self-hosting on Hetzner? How to seperate db from the rest of the backend etc.?

**Reply by aantti** [reply]
Yes, I'm testing https://supabase.com/docs/guides/self-hosting/docker

> How to seperate db from the rest of the backend etc.?

If you meant "how to move Supabase Postgres to a separate instance" - you can carve out and move the db service from `docker-compose.yml`, then change the configuration in `.env` to point to a different host. If it was about "how to use any arbitrary Postgres" - that's a bit harder, and will not be 100% compatible (esp., regarding the extensions like pg_net).

Not for Hetzner, but, as an example of a fully separate services configuration - https://github.com/BenIsenstein/pgonrails (via @BenIsenstein)

**Reply by aantti** [reply]
There's also https://github.com/supabase-community/supabase-kubernetes/ that should work on Hetzner, I'm guessing.
